gobblin-yarn/src/main/java/gobblin/yarn/GobblinApplicationMaster.java

```diff
@@ -0,0 +1,453 @@
+      LOGGER.info("Adding YarnContainerSecurityManager since security is enabled");
+      services.add(new YarnContainerSecurityManager(config, fs, this.eventBus));
+    }
+    services.add(
+        new YarnService(config, applicationName, applicationAttemptIdId.getApplicationId(), fs, this.eventBus,
+            Strings.nullToEmpty(config.getString(ConfigurationConstants.CONTAINER_JVM_ARGS_KEY))));
+    services.add(
+        new GobblinHelixJobScheduler(YarnHelixUtils.configToProperties(config), this.helixManager, this.eventBus,
+            appWorkDir));
+    services.add(new JobConfigurationManager(this.eventBus,
+        config.hasPath(ConfigurationConstants.JOB_CONF_PACKAGE_PATH_KEY) ? Optional
+            .of(config.getString(ConfigurationConstants.JOB_CONF_PACKAGE_PATH_KEY)) : Optional.<String>absent()));
+
+    this.serviceManager = new ServiceManager(services);
+  }
+
+  /**
+   * Start the ApplicationMaster.
+   */
+  public void start() {
+    LOGGER.info("Starting the Gobblin Yarn ApplicationMaster");
+
+    // Add a shutdown hook so the task scheduler gets properly shutdown
+    addShutdownHook();
+
+    try {
+      this.helixManager.connect();
+      this.helixManager.addLiveInstanceChangeListener(new GobblinLiveInstanceChangeListener());
+      this.helixManager.getMessagingService().registerMessageHandlerFactory(
+          Message.MessageType.SHUTDOWN.toString(), new ControllerShutdownMessageHandlerFactory());
+      this.helixManager.getMessagingService().registerMessageHandlerFactory(
+          Message.MessageType.USER_DEFINE_MSG.toString(), new ControllerUserDefinedMessageHandlerFactory()
+      );
+    } catch (Exception e) {
+      LOGGER.error("HelixManager failed to connect", e);
+      throw Throwables.propagate(e);
+    }
+
+    // Register JVM metrics to collect and report
+    registerJvmMetrics();
+
+    // Start metric reporting
+    this.jmxReporter.start();
+
+    // Start all the services running in the ApplicationMaster
+    this.serviceManager.startAsync();
+    this.serviceManager.awaitHealthy();
+  }
+
+  /**
+   * Stop the ApplicationMaster.
+   */
+  public synchronized void stop() {
+    if (this.isStopped || this.stopInProgress) {
+      return;
+    }
+
+    this.stopInProgress = true;
+
+    LOGGER.info("Stopping the Gobblin Yarn ApplicationMaster");
+
+    // Send a shutdown request to the containers as a second guard in case Yarn could not stop the containers
+    sendShutdownRequest();
+
+    try {
+      // Give the services 5 minutes to stop to ensure that we are responsive to shutdown requests
+      this.serviceManager.stopAsync().awaitStopped(5, TimeUnit.MINUTES);
+    } catch (TimeoutException te) {
+      LOGGER.error("Timeout in stopping the service manager", te);
+    } finally {
+      // Stop metric reporting
+      this.jmxReporter.stop();
+
+      if (this.helixManager.isConnected()) {
+        this.helixManager.disconnect();
+      }
+    }
+
+    this.isStopped = true;
```

Comment:
Yes, this is useful to prevent `stop` from being called more than once, which may happen if the AM receives a shutdown message and for some reason the shutdown hook is called due to some reason like running out of virtual memory limit.